py/sentry_streams/adapters/stream_adapter.py

```diff
@@ -19,6 +19,10 @@ def source(self, step: Step) -> Any:
     def sink(self, step: Step, stream: Any) -> Any:
         raise NotImplementedError
 
+    @abstractmethod
+    def map(self, step: Step, stream: Any) -> Any:
+        raise NotImplementedError
```

Comment:
I think we can be more specific with the Step type.

As each step type is translated with its own specific method we can make the adapter take `step: Map` and assert the type is right in the RuntimeTranslator:
```
elif step_type is StepType.MAP:
            return self.adapter.map(step, stream)
```

The rationale for this is that we can keep the logic that manages dynamic types in the same place at least.
In the current architecture the translator cannot request a specific type of Step by definition, so it has to infer, from the type, the method to call. 
Though we can still keep the adapter strictly typed as each method only takes one type of step. This allows us to rely on assertions only in one place and take advantage of typing as much as we can.